### ISSUE TITLE
More `const` functions

### ISSUE THREAD

**Issue by TH3-S4LM0N** [bevyengine/bevy#7557]
There are a lot of functions for building things that aren't declared as constant, even though they are constant.
The example that gets me the most is `bevy_render::mesh::shape::Box::new` isn't a `const` even though it's literally:
```rust
pub fn new(x_length: f32, y_length: f32, z_length: f32) -> Box {
        Box {
            max_x: x_length / 2.0,
            min_x: -x_length / 2.0,
            max_y: y_length / 2.0,
            min_y: -y_length / 2.0,
            max_z: z_length / 2.0,
            min_z: -z_length / 2.0,
        }
}
 ```
 and really has no reason not to be `const`. This is just an example, but there are tons of functions that should be constant.
 
 In the end this is just a suggestion, bevy is already huge so it might be unrealistic to go through it all again and figure out what can and can't be `const`, I'm just trying to bring attention to it.

**Comment by alice-i-cecile** [comment]
This is a good choice. Ideally this is tackled incrementally, maybe crate by crate?

**Comment by MinerSebas** [comment]
While I agree that more stuff should be const, that function is actually an example where it is still impossible in todays Rust:
```
error[E0658]: floating point arithmetic is not allowed in constant functions
  --> crates/bevy_render/src/mesh/shape/mod.rs:44:20
   |
44 |             max_x: x_length / 2.0,
   |                    ^^^^^^^^^^^^^^
   |
   = note: see issue #57241 <https://github.com/rust-lang/rust/issues/57241> for more information

```

**Comment by TH3-S4LM0N** [comment]
> Ideally this is tackled incrementally, maybe crate by crate?
This was my general idea, although it seems (to me) that it would take multiple people. Originally I was thinking about doing myself but realized the size of bevy and that it's unrealistic for me to just sit down and look over it all.
tldr this probably needs some kind of delegation.

**Comment by KenCox94** [comment]
I'd like to help out on this!

**Comment by alice-i-cecile** [comment]
@KenCox94 pick a small crate, comment that you're tackling it here, and constify what you can :) Open a PR for it and link it to this PR.

**Comment by valaphee** [comment]
Wouldn't constify everything, as it is not needed/has no effect in most scenarios

For example, Box::new(x_length: f32, y_length: f32, z_length: f32) will generate
```llvm
define void @bevy_render::mesh::shape::Box::new(ptr noalias nocapture noundef writeonly sret(%"mesh::shape::Box"
start:
%1 = insertelement <4 x float> poison, float %x_length, i64 0
%2 = insertelement <4 x float> %1, float %y_length, i64 1
%shuffle = shufflevector <4 x float> %2, <4 x float> poison, <4 x i32> <i32 0, i32 0, i32 1, i32 1>
%3 = fmul <4 x float> %shuffle, <float -5.000000e-01, float 5.000000e-01, float -5.000000e-01, float 5.000000e-01>
%4 = insertelement <2 x float> poison, float %z_length, i64 0
%5 = shufflevector <2 x float> %4, <2 x float> poison, <2 x i32> zeroinitializer
%6 = fmul <2 x float> %5, <float -5.000000e-01, float 5.000000e-01>
store <4 x float> %3, ptr %0, align 4
%7 = getelementptr inbounds %"mesh::shape::Box"
store <2 x float> %6, ptr %7, align 4
ret void
}
```

But when this method is being used, it will look like
```llvm
define void @bevy_render::mesh::shape::Box::use_box(ptr noalias nocapture noundef writeonly sret(%"mesh::shape::Box"
start:
store <4 x float> <float -5.000000e-01, float 5.000000e-01, float -1.000000e+00, float 1.000000e+00>, ptr %boxci, align 4, !alias.scope !80311
%0 = getelementptr inbounds %"mesh::shape::Box"
store <2 x float> <float -1.500000e+00, float 1.500000e+00>, ptr %0, align 4, !alias.scope !80311
ret void
}
```
And as you can see, it automatically recognized that it is const

**Comment by KenCox94** [comment]
@alice-i-cecile im assuming I should:

1. add const where it is truly "needed" 
2. looking for builder functions

I see a few property functions like `is_paused` that simply return an attribute value but I dont want to go around frivolously marking const because I can. 

Are my assumptions correct? 
Would you like for a "if it can const, const it" approach.

**Comment by alice-i-cecile** [comment]
Let's start with the former for now; I know @cart has been opposed to adding things like const and must_use  super aggresively in the past.

**Comment by mockersf** [comment]
Please be very mindful when adding `const`, most of the time it doesn't make sense to add it just because you can

**Comment by alice-i-cecile** [comment]
Agreed, we should only add const when a) it's clearly useful and b) we're confident that methods will continue to remain const.

We're also *really* missing const float arithmetic. There's a larger number of builder methods in geometry and UI that would really like this.

**Comment by james7132** [comment]
Tentatively closing this out since this issue isn't really actionable and we've been making APIs `const` as needed when they come into via PRs.
